drivers: video: Add the video shell

josuah opened the pull request:
Dependencies:

- #82158 

Replaces:

- #82393

Now that we have a nice shell control querying API, it is possible to tab-complete in-depth every control value.

```
video - Video driver commands
Subcommands:
  start    : Start a video device and its sources
             Usage: video start <device>
  stop     : Stop a video device and its sources
             Usage: video stop <device>
  capture  : Capture a given number of frames from a device
             Usage: video capture <device> <num-frames>
  format   : Query or set the video format of a device
             Usage: video format <device> <ep> [<fourcc> <width>x<height>]
  frmival  : Query or set the video frame rate/interval of a device
             Usage: video frmival <device> <ep> [<n>fps|<n>ms|<n>us]
  ctrl     : Query or set video controls of a device
             Usage: video ctrl <device> [<ctrl> <value>]
```

To test on `native_sim` with `VIDEO_SW_GENERATOR`:

```
west build -t run -b native_sim/native/64 samples/drivers/video/capture -- -DCONFIG_VIDEO_SHELL=y
```

On another terminal, using your favorite serial console client, and likely `COM#` on Windows instead of `/dev/pts/3`:

```
picocom -b 115200 /dev/pts/3
```

![](https://private-user-images.githubusercontent.com/55351021/433174690-c3637ec5-ee39-49f5-9660-75b59da9cb72.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ1OTUwMjksIm5iZiI6MTc0NDU5NDcyOSwicGF0aCI6Ii81NTM1MTAyMS80MzMxNzQ2OTAtYzM2MzdlYzUtZWUzOS00OWY1LTk2NjAtNzViNTlkYTljYjcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE0VDAxMzg0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwOTNiY2JlMmFkMmVjZTA2OTZiZTAxNzdlMTNiMDc0YzhmZWQ1YTA1MjMxN2ZjZDc2ZTgwYTQ1MGVhMmNmNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qOLJ2C8s8-Jn-Vri-vgT_FfUY7o4DbfYRlIAQKiJM4k)

josuah committed:
drivers: video: shell: introduce the video shell

Introduce the video shell and implement some video shell commands.
Make use of the various querying API to implement tab-completion, and
validiate the data, as well as convert string names into integers.
Commands provided: frmival, format, ctrl, start, stop, capture

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Because of the limitation of the shell API, a hack was used using `LISTIFY()` to generate completion functions that have an extra argument (i.e. the device number), so that the tab completion results are aware of the beginning of the line.

This is not expected to take more RAM, only more ROM/Flash.

josuah commented:
Force-push:

- Add the `capture` command mimicking the behavior of `yavta -c<num-frames> <device>` with `video capture <device> <num-frames>`.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -697,6 +697,20 @@ struct video_ctrl_query;
+ *
+ * For instance, "White Balance, Automatic" becomes "white_balance_automatic".
+ *
+ * @brief src Control name as generated by @ref video_query_ctrl
+ * @brief dst Destination buffer where the converted name is stored.
+ * @brief dst_size Maximum size of @p dst including the NUL string terminator.
```
Silly me!

josuah commented:
My bad, here we go again. No complain from:
```
cmake --build _build --target doxygen
scripts/ci/check_compliance.py -n -c origin/main..HEAD -e ClangFormat
```

josuah commented on drivers/video/video_ctrls.c:
```diff
@@ -421,6 +422,23 @@ static inline const char *video_get_ctrl_name(uint32_t id)
+	size_t i;
+	bool add_underscore = false;
+
+	for (i = 0; i + 1 < dst_size && src[i] != '\0'; i++) {
+		if (isalnum(src[i])) {
+			dst[i] = tolower(src[i]);
+			add_underscore = true;
+		} else if (add_underscore && src[i + 1] != '\0') {
+			dst[i] = '_';
+			add_underscore = false;
+		}
+	}
+	dst[i] = '\0';
```
good opportunity to add unit tests...

kartben commented on tests/drivers/video/ctrl/src/main.c:
```diff
@@ -0,0 +1,32 @@
+		{"nothing exceptional", "nothing_exceptional"},
+		{"UPPERCASE", "uppercase"},
+		{"._!!_###_!!_.", ""},
+		{"  _  leading special characters", "leading_special_characters"},
+		{"trailing special characters!@#$%^", "trailing_special_characters_"},
```
so trailing special characters are folded into 1 underscore and leading special characters end up being discarded? the lack of symmetry seems odd, maybe?

josuah replied:
> the lack of symmetry seems odd, maybe?

I agree. This was assuming a control would never be named i.e. `Misc +`.
Latest push also covers this edge case instead of asserting a mis-handling.

josuah commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1012 @@
+	SHELL_DYNAMIC_CMD_CREATE(dsub_##name##n, complete_##name##n)
+
+/* Maximum size for a video control name, be it standard one or custom device-specific one */
+#define VIDEO_SHELL_NAME_MAX_SIZE 40
```
I opted for a more global `CONFIG_VIDEO_CTRL_NAME_MAX_SIZE`.

josuah commented on drivers/video/video_ctrls.c:
```diff
@@ -462,6 +480,7 @@ void video_print_ctrl(const struct device *const dev, const struct video_ctrl_qu
 	uint8_t i = 0;
 	const char *type = NULL;
 	char typebuf[8];
+	char namebuf[40];
```
Applied, good point.

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -697,6 +697,20 @@ struct video_ctrl_query;
+ * @param dst Destination buffer where the converted name is stored.
+ * @param dst_size Maximum size of @p dst including the NUL string terminator.
+ */
+void video_convert_ctrl_name(char const *src, char *dst, size_t dst_size);
```
This is actually a convenient function to add underscores between words of a given input string. Its purpose is to give users a bit more convenience when using the tool, to "tab_complete" the ctrl name instead of typing the whole one ?

If so, by looking at its functionality (not a video-related function like set_fmt, set_stream, etc.) and its usage, I don't think it should be put in `video.h` as a public video API (at this point, need a release note, unittest, etc.) to be used by both drivers and applications. I suggest to put it internally in the `video-ctrls.c`

josuah replied:
> give users a bit more convenience when using the video shell tool

Yes exactly this, having spaces in the control name would also give a different number of arguments.
This also permitted to keep names matching with the `v4l2-ctl`:

```
v4l2-ctl --device /dev/video0 --set-ctrl white_balance_automatic=0
```

josuah replied:
This is now done :)

ngphibang commented on drivers/video/video_ctrls.c:
```diff
@@ -492,16 +516,18 @@ void video_print_ctrl(const struct device *const dev, const struct video_ctrl_qu
 
 	video_get_ctrl(dev, &vc);
 
+	video_convert_ctrl_name(cq->name, namebuf, sizeof(namebuf));
```
Is there a way to do this inside the video shell as `v4l2-ctl` did (as last time you saw the link) because this is actually for an additional convenience (tab-complete) when using that tool which should not interfere with the main function ?

If we do it here, we force everyone to always print the control names with underscores and don't use at all their original names, which is too video shell oriented (and if we do so, another question raise "why not just make the names with underscore right at the beginning?")

ngphibang replied:
If it's not easy, maybe we can cover this with `if (IS_ENABLED(CONFIG_VIDEO_SHELL) `?

josuah replied:
> Linux has reason not to do this because it wants to keep a human readable format which is clear and descriptive

That could be required for a small HTML-based or LVGL-based user interface, like a "camera controls" menu.

Currently the cost is not too big for the Zephyr shell: only a single string held in memory at a time (in a global [`video_shell_ctrl_name`](https://github.com/zephyrproject-rtos/zephyr/pull/88566/files#diff-72e4f24fa1184b77a0f306ac62c7133ed315c586010b9975ed5f4ad90485a3a2R39) variable) though it is not thread-safe.

> we force everyone to always print the control names with underscores
> we can cover this with if (IS_ENABLED(CONFIG_VIDEO_SHELL)

It looks like we can modify `cq->name` before passing it to `video_print_ctrl()`.

ngphibang reviewed (commented):
The shell tool is useful to help to do something similar to the video capture sample whereas we have the flexibility to change things at runtime (format, ctrl, devices, etc.).

While I see commands like capture, set_ctrl, etc. are helpful, commands like stop / start seems not have its real-world usage, just for demo-ing purpose (Maybe I miss something, to capture, we don't need to do start / stop ?) ?

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,963 @@
+	struct video_ctrl_query cq = {0};
+	struct video_control ctrl;
+	char menu_name[CONFIG_VIDEO_CTRL_NAME_MAX_SIZE];
+	char *arg0 = argv[0];
```
Should be `argv[1]` , i.e. ctrl value, otherwise it won't work

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,963 @@
+			return -ENOENT;
+		}
+
+		video_convert_ctrl_name(name, ctrl_name, sizeof(ctrl_name));
```
Should be `video_convert_ctrl_name(cq->name, ctrl_name, sizeof(ctrl_name));` to add underscore to the cq, not the "name" who has already underscore.

ngphibang commented on drivers/video/video_ctrls.h:
```diff
@@ -68,4 +68,6 @@ void video_cluster_ctrl(struct video_ctrl *ctrls, uint8_t sz);
 
 void video_auto_cluster_ctrl(struct video_ctrl *ctrls, uint8_t sz, bool set_volatile);
 
+void video_convert_ctrl_name(char const *src, char *dst, size_t dst_size);
```
As being said, it is rather an utility function to add underscores to a string, not specific to the video ctrl. Could we rename it to something like "add_underscore()" and move it to where it is used the `video_shell.c` ?

josuah replied:
I understand now, thank you!

ngphibang reviewed (commented):
So far, I just tested only the capture command (ít works fine) and the `set_ctrl` command and see it works wrongly, mainly just because some mistakes (I pointed out in other comments). Could you test the rest ?

![Screenshot from 2025-05-08 01-28-07](https://github.com/user-attachments/assets/29e386fe-24bc-4bae-a876-7ad473e27439)


![Screenshot from 2025-05-08 01-28-47](https://github.com/user-attachments/assets/bbe02258-43a5-4518-b7df-975a8227f7d2)

ngphibang commented:
> So far, I used the start/stop commands to pause and resume a feed started by the application to see what this does on the device. Like a missing "play/pause" button for the application.
> 
> Another use is abusing the video_stream_start() as a "trigger one frame capture" mode in the meantime that a dedicated API is there.
> 
> Another is starting and stopping a device while a logic analyzer/oscilloscope is plugged to see what happens.

Thanks, I see their usages now. They should be added then (sorry). So far, just recheck and see yavta also includes start / stop command.

ngphibang commented:
One point I don't understand how you get this work on a real board like Arduno Nicla Vision ? Because when built with the video capture sample, the camera is used by the sample and not availble for the video shell. I had to comment out all the code in the sample

` west build -t run -b mimxrt1170_evk/mimxrt1176/cm7 samples/drivers/video/capture -- -DSHIELD='nxp_btb44_ov5640 rk055hdmipi4ma0' -DCONFIG_VIDEO_SHELL=y`

ngphibang commented:
Another point is that the shell displays all the video devices including : sensor, receiver, bridge devices whereas we should expose only the bridge / main device of a pipeline. That's why https://github.com/zephyrproject-rtos/zephyr/pull/88182 should come into play.

josuah commented:
> One point I don't understand how you get this work on a real board like Arduno Nicla Vision?

The Nicla Vision did work as it only has vflip/hflip, not affected by the bug.

I did not find working hardware at hand with integer control types (`mimxrt1170_evk` coming in the mail soon), but we are porting our SDK to latest Zephyr then we can use it to test with the IMX219.

josuah commented:
Much better now:

![scrot_20250508_175452_1920x1080](https://github.com/user-attachments/assets/9aa64f9f-fe7a-4691-a186-d2a238195b14)
Two cameras, tuning the exposure and gain value of the top one. As noisy as IMX219 gets under high-gain!

josuah commented:
> One point I don't understand how you get this work when building with the video capture sample ? Because the camera is then used by the sample and not availble for the video shell. I had to comment out all the code in the sample. It seems I do it wrongly ?

I had to comment out all the code in the sample as well. I forgot to mention it.
Should a `samples/drivers/video/shell/` be added? Likely without any `src/main.c`.

josuah commented:
force-push:
- use explicit `default:` for switch as rerquested by SonarQube

josuah committed:
sample: drivers: video: capture: add support for the video shell

Modify the capture sample to only run the video shell when the user
enables it with -DCONFIG_VIDEO_SHELL=y, as an alternative way to trigger
captures.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force-push:

- Add a shell sample

I am not sure I am doing the device testing right.
<details><summary><code>west twister --device-testing -p native_sim -T samples/drivers/video/shell/</code></summary>
<pre><code>
INFO    - Using Ninja..
INFO    - Zephyr version: v4.1.0-3534-ga90a7dc7e790
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/testplan.json

Device testing on:

| Platform   | ID   | Serial device   |
|------------|------|-----------------|

INFO    - JOBS: 1
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue

INFO    - 1 test scenarios (1 configurations) selected, 1 configurations filtered (1 by static filter, 0 at runtime).
INFO    - 0 of 0 executed test configurations passed (0.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 2.59 seconds.
INFO    - 0 of 0 executed test cases passed (0.00%) on 0 out of total 1042 platforms (0.00%).
INFO    - 0 test configurations executed on platforms, 0 test configurations were only built.

Hardware distribution summary:

| Board   | ID   | Counter   | Failures   |
|---------|------|-----------|------------|
INFO    - Saving reports...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
</code></pre>
</details>

ngphibang commented:
> One point I don't understand how you get this work when building with the video capture sample ? Because the camera is then used by the sample and not availble for the video shell. I had to comment out all the code in the sample. It seems I do it wrongly ?
> 
> I had to comment out all the code in the sample as well. I forgot to mention it.
> Should a samples/drivers/video/shell/ be added? Likely without any src/main.c.
> 
> Yes please! It needs the main I think, but will be empty

Having an empty main is the correct way. But I suspect if having an own sample for video shell is the right way to do.
At the moment, we have 3 video samples which do (very) similar things:

- video capture
- video capture with lvgl
- video shell

where code is nearly repeated for supported boards (overlay, configs, README) and we also take efforts for maintenance. As we can see video capture sample got much more supports than video capture with lvgl.

Would it be better to add "if (IS_ENABLED(CONFIG_VIDEO_SHELL)" in the `main.c` of the video capture sample to have an empty main there ?

josuah commented:
> Would it be better to add "if (IS_ENABLED(CONFIG_VIDEO_SHELL)" in the main.c of the video capture sample to have an empty main there ?

Yes I picked that method now, and added another test for it, it should run on CI using the `native_sim`:
<details>
<summary><code>twister-out/native_sim_native_64/host/samples/drivers/video/capture/sample.video.capture.shell/handler.log</code></summary>
<pre><code>uart connected to pseudotty: /dev/pts/3
*** Booting Zephyr OS build v4.1.0-3534-g30e3c515f796 ***
[00:00:00.000,000] <inf> main: Video device: VIDEO_SW_GENERATOR
[00:00:00.000,000] <inf> main: - Capabilities:
[00:00:00.000,000] <inf> main:   RGBP width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   BX24 width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main: - Video format: RGBP 320x160
[00:00:00.000,000] <inf> main: - Default frame rate : 30.000000 fps
[00:00:00.000,000] <inf> main: - Supported frame intervals for the default format:
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main: - Supported controls:
[00:00:00.000,000] <inf> video_ctrls:                  Horizontal Flip 0x00980914 (bool)   (flags=0x00) : min=0 max=1 step=1 default=0 value=0
[00:00:00.000,000] <inf> main: Letting the user control the device with the video shell
</pre></code>
</details>

josuah commented:
Force-push:

- Remove the shell sample
- Modify the capture sample to allow the shell with `if (IS_ENABLED(CONFIG_VIDEO_SHELL))`

josuah commented:
Force-push:

- Rebase on main, fix conflict

ngphibang commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -199,10 +199,16 @@ int main(void)
-	ctrl.id = VIDEO_CID_TEST_PATTERN;
-	video_set_ctrl(video_dev, &ctrl);
-#endif
+	if (IS_ENABLED(CONFIG_TEST)) {
+		ctrl.id = VIDEO_CID_TEST_PATTERN;
+		video_set_ctrl(video_dev, &ctrl);
+	}
```
This is unrelated to the video shell change. While it is better to use `if` instead of `#ifdef` here, this should be in a separate commit so that people will not be confused when reading commit history.

josuah replied:
Good point, moved and also converted the other in the process.

ngphibang commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -199,10 +199,16 @@ int main(void)
+	}
+
+	/* When the video shell is enabled, do not run the capture loop */
+	if (IS_ENABLED(CONFIG_VIDEO_SHELL)) {
+		LOG_INF("Letting the user control the device with the video shell");
+		return 0;
+	}
```
I think we should return as soon as possible, right at the beginning of the main function.

- By doing this here, the sample already reserved some video buffers for nothing:

`struct video_buffer *buffers[CONFIG_VIDEO_BUFFER_POOL_NUM_MAX], *vbuf;`

Users can be very fast (or if something is stuck in the sample) doing something already in the shell before this point, so we waste a significant amount of memory and not enough memory for the shell.

- In fact, the idea is we can use video shell not only on the video device in the sample but also for other video devices (e.g. sw generator). And all the logs information in the sample up to this point (format, framerate, controls) are not needed as user can get them with the shell commands ?

josuah replied:
Good point, the video shell already offers a way to print all of the early main content, and the main loop as well.

I moved the `if (IS_ENABLED(CONFIG_VIDEO_SHELL))` all the way up. Easy to see that nothing is required for the shell to work that way.

josuah replied:
Currently, all the sample main loop does is to just log the buffer fields and stays at that:
```c
                LOG_DBG("Got frame %u! size: %u; timestamp %u ms", frame++, vbuf->bytesused,
                       vbuf->timestamp);
```

And the video shell actually does not do it. Should I add buffer timestamps to the shell `capture` command?
The basic logging of the shell sample can still be useful for troubleshooting when "debug" logs are enabled.

https://github.com/zephyrproject-rtos/zephyr/compare/e992d709d11230d26ea32e5844d289441eda2eef..3e707557fb8eb7486310b024f5c663862d9ce3d5

ngphibang replied:
Yes, I think

josuah committed:
samples: drivers: video: capture: convert #ifdef into if()

Convert #ifdef CONFIG_TEST into if (IS_ENABLED(CONFIG_TEST)) to improve
readability.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1042 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <ctype.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/video.h>
+#include <zephyr/drivers/video-controls.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/shell/shell.h>
```
Thank you, I keep forgetting...

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1053 @@
+				  enum video_endpoint_id ep, size_t argc, char **argv)
+{
+	struct video_format fmt = {0};
+	char *arg1 = argv[1];
```
idem

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1053 @@
+	struct video_ctrl_query cq = {0};
+	struct video_control ctrl;
+	char menu_name[CONFIG_VIDEO_SHELL_CTRL_NAME_SIZE];
+	char *arg1 = argv[1];
```
idem

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1053 @@
+		    argv[0], ctrl.id, argv[1], ctrl.val64);
+
+	if (cq.type != VIDEO_CTRL_TYPE_INTEGER64) {
+		ctrl.val = ctrl.val64;
```
This is not needed as val and val64 are in the same union so they point to the same memory address ?

josuah replied:
I have no problem with removing it, but is that also a no-op for big-endian architectures?

ngphibang replied:
Ok, I see now the concern for BE / LE. In fact, it's related to the implementation earlier, if val and val64 are used "normally", this is not a concern (?). It does not matter to me, no need to change if you prefer to do it this way.

josuah replied:
I went for something with 1 more line of code but more straightforward.
In hope that makes sense.
Thank you for the feedback.

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1053 @@
+	ctrl.id = cq.id;
+
+	switch (cq.type) {
+	case VIDEO_CTRL_TYPE_BOOLEAN:
+		if (strcmp(argv[1], "on") == 0) {
+			ctrl.val64 = true;
+		} else if (strcmp(argv[1], "off") == 0) {
+			ctrl.val64 = false;
+		} else {
+			shell_error(sh, "Video control value must be 'on' or 'off', not '%s'",
+				    argv[1]);
+			return -EINVAL;
+		}
```
Is it simpler to just use 1 and 0 interger instead of "on" and "off" string as v4l2-ctl ? Query-ctrl command can show the min, max, def, cur values so users will not confuse.

josuah replied:
Having "on" and "off" permitted the user to discover the type is bool with just 'tab'.
I can offer completion candidates for only `0` and `1` then?

ngphibang replied:
I see, so it's ok as currently then.

josuah replied:
I also added `"0"` and `"1"` as possible input values even though not tab completed. I can also add them as tab completion candidates, very easy.

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1053 @@
+		for (i = 0; cq.menu[i] != NULL; i++) {
+			video_shell_convert_ctrl_name(cq.menu[i], menu_name, sizeof(menu_name));
+			if (strcmp(menu_name, argv[1]) == 0) {
+				ctrl.val64 = i;
```
type menu value is integer, so it should be `ctrl.val` (though they can be used interchangeably) ?

josuah replied:
I did set everything to `ctrl.val64` so that `strtoull()` could be used for every type:

```c
		ctrl.val64 = strtoull(arg1, &arg1, 10);
```

But that is confusing, so let's get back to something more readable.

ngphibang replied:
It's not confusing, I see the intention, just a bit unusual where I see `ctrl.val = ctrl.val64` as commented above. Keep the implementation as is if you see it's more convenient.

ngphibang commented on drivers/video/video_shell.c:
```diff
@@ -0,0 +1,1053 @@
+/* Current video control, used for tab completion */
+static struct video_ctrl_query video_shell_cq;
+
+/* Current video endpoint, used for tab completion */
+static enum video_endpoint_id video_shell_ep = VIDEO_EP_NONE;
```
VIDEO_EP_ is dropped so this need to be rebased. Idem for others.

josuah replied:
Hurray!

ngphibang reviewed (commented):
VIDEO_EP_ is dropped but I see that the PR still uses it and passes CI (?). That means there is no built-only test cover the video shell. I don't know if we could add a built-only test for it ?

PS : It seems we just need to re-trigger the CI

josuah commented:
Force-push:

- Address [SCA warnings](https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=88566&id=zephyrproject-rtos_zephyr&open=AZa7vjYoCxoDuXFz8dPr)
- Set vbuf->type before dequeuing
- strtoull() -> strtoll()
- Move for a more straightforward way to assing ctrl.val/ctrl.val64
- Drop `VIDEO_EP_###`
- Allow `"0"` or `"1"` to be used instead of `"on"` or `"off"`
